cpp/src/slice2py/PythonUtil.cpp

```diff
@@ -2552,53 +2521,134 @@ namespace
-        if (parseStatus == EXIT_FAILURE)
-        {
+            if (parseStatus == EXIT_FAILURE)
+            {
+                status = EXIT_FAILURE;
+            }
+            else
+            {
+                if (dependencyGenerator)
+                {
+                    // Collect the dependencies of the unit.
+                    dependencyGenerator->addDependenciesFor(unit);
+                }
+
+                parseAllDocComments(unit, Slice::Python::pyLinkFormatter);
+                validatePythonMetadata(unit);
+
+                unit->visit(&packageVisitor);
+                unit->visit(&importVisitor);
+
+                if (compilationKind == PythonCompilationKind::All || compilationKind == PythonCompilationKind::Module)
+                {
+                    CodeVisitor codeVisitor{
+                        importVisitor.getRuntimeImports(),
+                        importVisitor.getTypingImports(),
+                        importVisitor.getAllImportNames()};
+                    unit->visit(&codeVisitor);
+
+                    const vector<PythonCodeFragment>& newFragments = codeVisitor.codeFragments();
+                    fragments.insert(fragments.end(), newFragments.begin(), newFragments.end());
+                }
+            }
             unit->destroy();
-            throw runtime_error("Failed to parse Slice file: " + fileName);
         }
+        catch (...)
+        {
+            if (preprocessor)
+            {
+                preprocessor->close();
+            }
 
-        parseAllDocComments(unit, Slice::Python::pyLinkFormatter);
-        validatePythonMetadata(unit);
+            if (unit)
+            {
+                unit->destroy();
+            }
+            throw;
+        }
+    }
+
+    if (compilationKind == PythonCompilationKind::All || compilationKind == PythonCompilationKind::Module)
+    {
+        // Write the imports for each code fragment.
+        ImportsMap runtimeImports = importVisitor.getRuntimeImports();
+        ImportsMap typingImports = importVisitor.getTypingImports();
 
-        unit->visit(&packageVisitor);
-        unit->visit(&importVisitor);
+        for (auto& fragment : fragments)
+        {
+            Python::BufferedOutput out;
+            writeHeader(out);
+            writeImports(runtimeImports[fragment.moduleName], typingImports[fragment.moduleName], out);
+            out << sp;
+            out << fragment.code;
 
-        CodeVisitor codeVisitor{
-            importVisitor.getRuntimeImports(),
-            importVisitor.getTypingImports(),
-            importVisitor.getAllImportNames()};
-        unit->visit(&codeVisitor);
+            fragment.code = out.str();
+        }
 
-        const vector<PythonCodeFragment>& newFragments = codeVisitor.codeFragments();
-        fragments.insert(fragments.end(), newFragments.begin(), newFragments.end());
-        unit->destroy();
+        fragments = sortCodeFragments(fragments, importVisitor.getRuntimeImports());
```

Comment:
sortCodeFragments is only required when call by loadSlice, I will add a param to let the method know if sorting is required.